crates/parcel/src/request_tracker/request.rs

```diff
@@ -1,67 +1,126 @@
+pub struct RunRequestContext {
+  parent_request_id: Option<u64>,
+  run_request_fn: RunRequestFn,
+  reporter: Arc<dyn ReporterPlugin + Send>,
+  cache: CacheRef,
+  file_system: FileSystemRef,
+  plugins: PluginsRef,
+  config_loader: ConfigLoaderRef,
 }
 
-impl<'a, T: Clone> RunRequestContext<'a, T> {
+impl RunRequestContext {
   pub(crate) fn new(
-    parent_request_hash: Option<u64>,
-    request_tracker: &'a mut RequestTracker<T>,
+    parent_request_id: Option<u64>,
+    run_request_fn: RunRequestFn,
+    reporter: Arc<dyn ReporterPlugin + Send>,
+    cache: CacheRef,
+    file_system: FileSystemRef,
+    plugins: PluginsRef,
+    config_loader: ConfigLoaderRef,
   ) -> Self {
     Self {
-      parent_request_hash,
-      request_tracker,
+      parent_request_id,
+      run_request_fn,
+      reporter,
+      cache,
+      file_system,
+      plugins,
+      config_loader,
     }
   }
 
   /// Report an event.
   pub fn report(&self, event: ReporterEvent) {
-    self.request_tracker.report(event);
+    self
+      .reporter
+      .report(&event)
+      .expect("TODO this should be handled?")
   }
 
-  /// Run a child request to the current request.
-  #[allow(unused)]
-  pub fn run_request(&mut self, request: &impl Request<T>) -> anyhow::Result<T> {
-    self
-      .request_tracker
-      .run_child_request(request, self.parent_request_hash)
+  /// Run a child request to the current request
+  pub fn queue_request(
+    &mut self,
+    request: impl Request,
+    tx: Sender<anyhow::Result<RequestResult>>,
+  ) -> anyhow::Result<()> {
+    let request: Box<dyn Request> = Box::new(request);
+    let message = RunRequestMessage {
+      request,
+      response_tx: Some(tx),
+      parent_request_id: self.parent_request_id,
+    };
+    (*self.run_request_fn)(message);
+    Ok(())
+  }
+
+  pub fn cache(&self) -> &CacheRef {
+    &self.cache
+  }
+
+  pub fn file_system(&self) -> &FileSystemRef {
+    &self.file_system
+  }
+
+  pub fn plugins(&self) -> &PluginsRef {
+    &self.plugins
+  }
+
+  pub fn config(&self) -> &ConfigLoaderRef {
+    &self.config_loader
   }
 }
 
 // We can type this properly
 pub type RunRequestError = anyhow::Error;
 
-pub trait Request<T: Clone>: DynHash {
+pub trait Request: DynHash + Send + Debug + 'static {
   fn id(&self) -> u64 {
     let mut hasher = parcel_core::hash::IdentifierHasher::default();
     std::any::type_name::<Self>().hash(&mut hasher);
     self.dyn_hash(&mut hasher);
     hasher.finish()
   }
 
-  fn run(&self, request_context: RunRequestContext<T>)
-    -> Result<RequestResult<T>, RunRequestError>;
+  fn run(
+    &self,
+    request_context: RunRequestContext,
+  ) -> Result<ResultAndInvalidations, RunRequestError>;
 }
 
-dyn_hash::hash_trait_object!(<T: Clone> Request<T>);
+dyn_hash::hash_trait_object!(Request);
 
-#[derive(Debug, PartialEq)]
-pub struct RequestResult<Req> {
-  pub result: Req,
+#[derive(Debug, Clone, PartialEq)]
+pub struct ResultAndInvalidations {
```

Comment:
Maybe we should just get rid of this struct if we plan to automatically track invalidations through `fs`

Reply:
Totally and this was really just porting what we already have. I agree we should rethink some of these types.

Reply:
I don't have a good answer, most of the options I can think of aren't intuitive either. That's why I'm advocating for the autotracking. It's fine to leave as is.